### PR TITLE
Fix flaking unit test TestOperatorSync_NoOp

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
Problem was that the informer cache was not being ensured to be synced before running the tests.

```
E0705 08:14:52.443801    6362 sync.go:37] Error syncing cluster api controller: timed out waiting for the condition
--- FAIL: TestOperatorSync_NoOp (320.10s)
    --- FAIL: TestOperatorSync_NoOp/AWS (300.07s)
    	operator_test.go:165: 
    			Error Trace:	operator_test.go:165
    			Error:      	Received unexpected error:
    			            	timed out waiting for the condition
    			Test:       	TestOperatorSync_NoOp/AWS
    			Messages:   	unexpected sync failure
    	operator_test.go:166: 
```

### CODE REVIEW

**Comment by vikaschoudhary16** [comment]
```
--- FAIL: TestReconcileWithoutUnhealthyConditionsConfigMap (0.02s)
	machinehealthcheck_controller_test.go:262: Test case: recentlyUnhealthy. Expected RequeueAfter between: 5m0s and 5m1s, got: {true 4m59.992730885s}
```
This seems another flake.

**Comment by cynepco3hahue** [comment]
lgtm, I think we can also run `go optr.Run(2, stopCh)` instead of `sync` operation and it will fix the problem, but the current solution also works for me

**Comment by ingvagabund** [comment]
@vikaschoudhary16 thanks for the easy fix. 

> lgtm, I think we can also run go optr.Run(2, stopCh) instead of sync operation and it will fix the problem, but the current solution also works for me

As @cynepco3hahue, `go optr.Run(2, stopCh)` is the right way how to fix the case since it already waits for the changes and sets additional bits.

**Comment by ingvagabund** [comment]
/hold

please gimme me some time to review the PR

**Comment by enxebre** [comment]
the name of the test `TestOperatorSync_NoOp` seems to be outdated since the test is doing more and not even calling the sync function explicitely

File: pkg/operator/operator_test.go [hunk 1]
```diff
@@ -158,15 +135,10 @@ func TestOperatorSync_NoOp(t *testing.T) {
...
 			stopCh := make(<-chan struct{})
-			optr := newFakeOperator(nil, []runtime.Object{infra}, stopCh)
-			optr.imagesFile = imagesFilePath
-
-			err = optr.sync("test-key")
-			if !assert.NoError(t, err, "unexpected sync failure") {
-				t.Fatal()
-			}
+			optr := newFakeOperator(nil, []runtime.Object{newFeatureGate(v1.TechPreviewNoUpgrade), infra}, stopCh)
+			go optr.Run(2, stopCh)
```
Comment (frobware) [inline pkg/operator/operator_test.go]: The problem I have with this is the design of `Run()` itself.

That looks like:

```go
// Run runs the machine config operator.
func (optr *Operator) Run(workers int, stopCh <-chan struct{}) {
	defer utilruntime.HandleCrash()
	defer optr.queue.ShutDown()

	glog.Info("Starting Machine API Operator")
	defer glog.Info("Shutting down Machine API Operator")

	if !cache.WaitForCacheSync(stopCh,
		optr.deployListerSynced,
		optr.featureGateCacheSynced) {
		glog.Error("Failed to sync caches")
		return
	}
	glog.Info("Synced up caches")
	for i := 0; i < workers; i++ {
		go wait.Until(optr.worker, time.Second, stopCh)
	}

	<-stopCh
}
```

- The call to `WaitForCacheSync` can fail, yet we ignore the error.
- Callers of `operator.Run()` can silently fail
- Perhaps the cache in some runs is not sync'd and therefore flakes?

**Comment by frobware** [comment]
> Problem was that the informer cache was not being ensured to be synced before running the tests.

See: https://github.com/openshift/machine-api-operator/pull/343#discussion_r301477428

File: pkg/operator/operator_test.go (hunk 1 above)
Comment (vikaschoudhary16) [reply pkg/operator/operator_test.go]: Agree that optr.Run() should be returning the error ideally. Though we have not seen "Failed to sync caches" error log in any of those flakes.

This PR is for the flake mentioned in the description. Lets create another PR for returning the error.

Comment (frobware) [reply pkg/operator/operator_test.go]: > This PR is for the flake mentioned in the description. Lets create another PR for returning the error.

Please can we hold for the moment.

**Comment by frobware** [comment]
I see how it fixes the flake

/lgtm

File: pkg/operator/operator_test.go [hunk 2]
```diff
@@ -158,15 +135,10 @@ func TestOperatorSync_NoOp(t *testing.T) {
...
 
 			stopCh := make(<-chan struct{})
-			optr := newFakeOperator(nil, []runtime.Object{infra}, stopCh)
-			optr.imagesFile = imagesFilePath
-
-			err = optr.sync("test-key")
-			if !assert.NoError(t, err, "unexpected sync failure") {
-				t.Fatal()
-			}
+			optr := newFakeOperator(nil, []runtime.Object{newFeatureGate(v1.TechPreviewNoUpgrade), infra}, stopCh)
```
Comment (ingvagabund) [inline pkg/operator/operator_test.go]: Normally, I would be against passing `newFeatureGate(v1.TechPreviewNoUpgrade)` object to the fake informer as it is no needed for the test. Though, the operator processes an item only when a deployment or a feature gate object is created/updated/deleted. Without passing either of those, the operator workers will not process anything. So, passing the feature gate object is actually a trigger to render new deployment. In the real cluster there has to be at least one deployment available in order to trigger the processing loop.

**Pull request merged**